deployment: kill kube-controller-manager when adding a new node

NicolasT opened the pull request:
~We had a long-standing issue with cluster expansion, where after
creating and deploying a new node, no `kube-proxy` Pod got scheduled on
it, rendering the node unusable.~

~When investigating, one could observe an event on the `DaemonSet` where
the `DaemonSetController` from `kube-controller-manager` reported the
Pod was not schedulable, since the Node had not enough resources: 1 Pod
needed, 0 available.~

~This is indeed the case when we create a new Node object without
provisioning the node (yet), though once `kubelet` is running on the new
node, the available resources are properly propagated to the Node
object. However, DSC didn't reschedule a Pod.~

~At this point, it did however schedule a new Pod for the `calico-node`
`DaemonSet` :-/~

~After spending some time reading the `DaemonSetController` source-code,
and pondering the differences between the `calico-node` and `kube-proxy`
`DaemonSet`s, as well as running `kube-controller-manager` with a
slightly higher log-level, made one thing clear: where the `kube-proxy`
scheduling failed because of missing resources, the scheduling of
`calico-node` initially failed because of missing resources, as well as
a non-matching `nodeSelector` (for `beta.kubernetes.io/os: linux`).
Given the source-code of the controller, it turns out that, indeed, this
second reason (as it's called in said code) causes a different code-path
to be taken in the controller implementation.~

~Trying to emulate this 'fix' by simply adding a `nodeSelector` to the
`kube-proxy` `DaemonSet` indeed causes the issue to be solved, and a Pod
to be scheduled once the new node is properly provisioned.~

~A patch that went into Kubernetes 1.12.0 could've fixed this as well,
however in testing this seems not to be the case.~

~See: https://github.com/kubernetes/kubernetes/commit/c42439c6765589b03fd4f79b75e8f30b1ec8fcfc#diff-3af9d435092399b4bf3ef61e0387eb36
See: https://github.com/kubernetes/kubernetes/pull/67337~

Previous approach didn't work in the end. Now resorting to killing/restarting `kube-controller-manager` when deploying a new node.

bert-e commented:
# Hello nicolast,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/kube-proxy-daemonset` into
`development/2.0`
* [w/2.1/bugfix/kube-proxy-daemonset](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/kube-proxy-daemonset) will be merged into `development/2.1` 
* [w/2.2/bugfix/kube-proxy-daemonset](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/kube-proxy-daemonset) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

NicolasT commented:
@bert-e approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf










*The following options are set:* **approve**

NicolasT commented:
So, looks like this is *not* a definitive fix anyway :disappointed: 

New approach: `kube-controller-manager` :gun: everywhere whenever a node is added to the cluster.

NicolasT commented:
@bert-e wait

NicolasT committed:
salt: import fixed `orchestrate_show_sls`

In Salt 2018.3 and 2019.2, the `state.orchestrate_show_sls` runner is
broken. There's a patch merged in the current development version,
however.

This PR adds the fixed version to the `metalk8s_saltutil` runner module
for now.

See: https://github.com/saltstack/salt/issues/50556
See: https://github.com/saltstack/salt/pull/51465
See: https://github.com/saltstack/salt/issues/51951

NicolasT committed:
salt: add `metalk8s.minions_by_role` utility

NicolasT committed:
salt: kill all `kube-controller-manager`s when deploying a new node

We had a long-standing issue with cluster expansion, where after
creating and deploying a new node, no `kube-proxy` Pod got
scheduled on it, rendering the node unusable.

When investigating, one could observe an event on the `DaemonSet` where
the `DaemonSetController` from `kube-controller-manager` reported the
Pod was not schedulable, since the Node had not enough resources: 1 Pod
needed, 0 available.

This is indeed the case when we create a new Node object without
provisioning the node (yet), though once `kubelet` is running on the new
node, the available resources are properly propagated to the Node
object. However, DSC didn't reschedule a Pod.

At this point, it did however schedule a new Pod for the `calico-node`
`DaemonSet` :-/ However, on some occasions it did not, for unknown
reasons.

A patch that went into Kubernetes 1.12.0 could've fixed this as well,
however in testing this seems not to be the case.

To work-around this issue, we now kill all instances of
`kube-controller-manager` on all `master` nodes such that after restart,
the controller would properly detect the new node and schedule any
missing Pods accordingly.

This is, of course, a very coarse-grained 'solution'. It's acceptible,
however (for now at least), since a cluster can withstand a KCM restart
gracefully.

See: https://github.com/kubernetes/kubernetes/commit/c42439c6765589b03fd4f79b75e8f30b1ec8fcfc#diff-3af9d435092399b4bf3ef61e0387eb36
See: https://github.com/kubernetes/kubernetes/pull/67337
See: #1028
See: https://github.com/scality/metalk8s/pull/1028

NicolasT commented:
@bert-e reset

NicolasT commented:
@bert-e status

NicolasT commented:
@bert-e status

bert-e commented:
# Status





*Status report is not available.*

bert-e commented:
# History mismatch



Merge commit #770e5f0b4914978b3d77ded9c3c12b2cb816f952 on the integration branch
`w/2.1/bugfix/kube-proxy-daemonset` is merging a branch which is neither the current
branch `bugfix/kube-proxy-daemonset` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `bugfix/kube-proxy-daemonset` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/kube-proxy-daemonset` into
`development/2.0`
* [w/2.1/bugfix/kube-proxy-daemonset](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/bugfix/kube-proxy-daemonset) will be merged into `development/2.1` 
* [w/2.2/bugfix/kube-proxy-daemonset](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/bugfix/kube-proxy-daemonset) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

sayf-eddine-scality reviewed (approved):
works as intended

NicolasT commented:
@bert-e status

bert-e commented:
# Status





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

NicolasT commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye nicolast.
